src/main/java/com/google/devtools/build/lib/cmdline/RepositoryMapping.java

```diff
@@ -40,6 +40,13 @@ public abstract class RepositoryMapping {
   /** Returns all the entries in this repo mapping. */
   public abstract ImmutableMap<String, RepositoryName> entries();
 
+  /**
+   * Returns inverse entries mapping for faster lookup.
+   * Although it would be sufficient to just iterate over entries(),
+   * having an inverse map allows us to perform lookups more efficiently.
+   */
+  public abstract ImmutableMap<RepositoryName, String> inverseEntries();
```

Comment:
this could be a lot of extra memory usage, since we have a lot more repo mapping objects than just the one for the main repo.

I'm wondering if we should just have a special "main repo mapping" class that has an additional cached inverse entries map.

Reply:
I just realized that this PR would probably also require us to update this SkyFunction: https://cs.opensource.google/bazel/bazel/+/master:src/main/java/com/google/devtools/build/lib/bazel/bzlmod/ModuleExtensionRepoMappingEntriesFunction.java;l=32?q=Repomappingen&sq=&ss=bazel

I agree that a special class for the root module would be better.

Reply:
This might be uglier but how about:
1) we make this private
2) we only instantiate it over some threshold otherwise we do linear search

then the memory will increase but only where we want it to increase because the lookups will be slow?

It can still be true that we have something that's not accessed often but now uses more memory.

Reply:
I think we actually just want the main repo mapping object to be different. I'm preparing a change for it, but I'm not sure how to actually test its impact. (other than some manual testing, that is)

Reply:
> It can still be true that we have something that's not accessed often but now uses more memory.

(the reason is exactly this)